sycl/include/CL/sycl/handler.hpp

```diff
@@ -729,6 +810,305 @@ class __SYCL_EXPORT handler {
+        // LocalReds[WGSize] accumulates last/odd elements when the step
+        // of tree-reduction loop is not even.
+        typename Reduction::binary_operation BOp;
+        size_t PrevStep = WGSize;
+        for (size_t CurStep = PrevStep >> 1; CurStep > 0; CurStep >>= 1) {
+          if (LID < CurStep)
+            LocalReds[LID] = BOp(LocalReds[LID], LocalReds[LID + CurStep]);
+          else if (LID == CurStep && (PrevStep & 0x1))
+            LocalReds[WGSize] = BOp(LocalReds[WGSize], LocalReds[PrevStep - 1]);
+          NDIt.barrier();
+          PrevStep = CurStep;
+        }
+
+        // Compute the the partial sum/reduction for the work-group.
+        if (LID == 0) {
+          auto GrID = NDIt.get_group_linear_id();
+          auto V = BOp(LocalReds[0], LocalReds[WGSize]);
+          Out.get_pointer().get()[GrID] =
+              IsUpdateOfUserAcc ? BOp(*(Out.get_pointer()), V) : V;
+        }
+      });
+    }
+  }
+
+  /// Implements a command group function that enqueues a kernel that does one
+  /// iteration of reduction of elements in each of work-groups.
+  /// This version uses tree-reduction algorithm to reduce elements in each
+  /// of work-groups. At the end of each work-groups the partial sum is written
+  /// to a global buffer.
+  ///
+  /// Briefly: aux kernel, tree-reduction, CUSTOM types/ops.
+  template <typename KernelName, typename KernelType, int Dims, class Reduction>
+  void reduAuxCGFunc(const nd_range<Dims> &Range, size_t NWorkItems,
+                     size_t KernelRun, Reduction &Redu) {
+    size_t WGSize = Range.get_local_range().size();
+    size_t NWorkGroups = Range.get_group_range().size();
+
+    // The last work-group may be not fully loaded with work, or the work group
+    // size may be not power of those. Those two cases considered inefficient
+    // as they require additional code and checks in the kernel.
+    bool IsUnderLoaded = NWorkGroups * WGSize != NWorkItems;
+    size_t InefficientCase = (IsUnderLoaded || (WGSize & (WGSize - 1))) ? 1 : 0;
+
+    bool IsUpdateOfUserAcc =
+        Reduction::accessor_mode == access::mode::read_write &&
+        NWorkGroups == 1;
+
+    // Use local memory to reduce elements in work-groups into 0-th element.
+    // If WGSize is not power of two, then WGSize+1 elements are allocated.
+    // The additional last element is used to catch reduce elements that
+    // could otherwise be lost in the tree-reduction algorithm.
+    auto LocalReds = Redu.getReadWriteLocalAcc(WGSize + InefficientCase, *this);
+
+    // Get read accessor to the buffer that was used as output
+    // in the previous kernel. After that create new output buffer if needed
+    // and get accessor to it (or use reduction's accessor if the kernel
+    // is the last one).
+    auto In = Redu.getReadAccToPreviousPartialReds(*this);
+    auto Out = Redu.getWriteAccForPartialReds(NWorkGroups, KernelRun, *this);
+
+    if (!InefficientCase) {
```

Comment:
May `if(EfficientCase)` be better than double negation from your point of view?

Reply:
I thought about that too. If it causes questions, it makes sense to fix it. Thank you.